### PR TITLE
Improve snapshot extensibility for NoSQL databases

### PR DESCRIPTION
**Pull request by fbiville**
## Impact

- [ ] Bug fix (non-breaking change which fixes expected existing functionality)
- [x] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

Pro companion PR: https://github.com/liquibase/liquibase-pro/pull/1605
Pro-tests companion PR: https://github.com/liquibase/liquibase-pro-tests/pull/1322
 
## Things to be aware of

* Fixes SnapshotGeneratorChain
* Improves Database.supports methods

Having a generic method as 
```
 default boolean supports(Class<? extends DatabaseObject> object) {
```

allows extensions (specially NoSql ones) to implement validations as 

```
supports(Table.class) && supports(Catalog.class) 
```

without changing core to add a new "supportsTables" method.

## Things to worry about
TODO

### CODE REVIEW

File: liquibase-standard/src/main/java/liquibase/snapshot/jvm/ViewSnapshotGenerator.java [hunk 1]
```diff
@@ -127,28 +127,32 @@ protected void addTo(DatabaseObject foundObject, DatabaseSnapshot snapshot) thro
...
-                    view.setName(row.getString("TABLE_NAME"));
-                    view.setSchema(new Schema(catalogAndSchema.getCatalogName(), catalogAndSchema.getSchemaName()));
-                    view.setRemarks(row.getString("REMARKS"));
-                    String definition = StringUtil.standardizeLineEndings(row.getString("OBJECT_BODY"));
-                    view.setDefinition(definition);
-                    if(database instanceof OracleDatabase) {
-                        view.setAttribute("editioning", "Y".equals(row.getString("EDITIONING_VIEW")));
-                    }
-                    schema.addDatabaseObject(view);
+        if (!(foundObject instanceof Schema)) {
```
Comment (fbiville) [inline liquibase-standard/src/main/java/liquibase/snapshot/jvm/ViewSnapshotGenerator.java:130]: just inverted the if to reduce indentation

File: liquibase-standard/src/main/java/liquibase/snapshot/jvm/ViewSnapshotGenerator.java [hunk 2]
```diff
@@ -127,28 +127,32 @@ protected void addTo(DatabaseObject foundObject, DatabaseSnapshot snapshot) thro
...
+            return;
+        }
+        Database database = snapshot.getDatabase();
+        if (!database.supportsViews()) {
```
Comment (fbiville) [inline liquibase-standard/src/main/java/liquibase/snapshot/jvm/ViewSnapshotGenerator.java:134]: ... and added that guard
that's all there is to this PR :)

**Review by filipelautert — APPROVED** [review]

File: liquibase-integration-tests/src/test/java/liquibase/dbtest/AbstractIntegrationTest.java [hunk 3]
```diff
@@ -202,11 +202,11 @@ protected void wipeDatabase() {
             String altSchema = testSystem.getAltSchema();
             String altCatalog = testSystem.getAltCatalog();
 
-            if (database.supportsSchemas()) {
+            if (database.supports(Schema.class)) {
```
Comment (fbiville) [inline liquibase-integration-tests/src/test/java/liquibase/dbtest/AbstractIntegrationTest.java:205]: I can move these changes to another PR, so this PR can remain small

File: liquibase-standard/src/main/java/liquibase/snapshot/SnapshotControl.java [hunk 4]
```diff
@@ -123,9 +123,9 @@ public String getSerializableFieldNamespace(String field) {
 
     private void setTypes(Set<Class<? extends DatabaseObject>> types, Database database) {
         this.types = new HashSet<>();
-        for (Class<? extends DatabaseObject> type : types) {
-            addType(type, database);
-        }
+        types.stream()
+                .filter(database::supports)
```
Comment (fbiville) [inline liquibase-standard/src/main/java/liquibase/snapshot/SnapshotControl.java:127]: this filter is new, and basically filters out objects the db does not support

File: liquibase-standard/src/main/java/liquibase/snapshot/SnapshotGeneratorChain.java [hunk 5]
```diff
@@ -40,34 +42,31 @@ public <T extends DatabaseObject> T snapshot(T example, DatabaseSnapshot snapsho
...
+            if (replacedGenerators.stream()
+                    .anyMatch(Predicate.isEqual(generator.getClass()))) {
+                continue;
+            }
+            T object = generator.snapshot(example, snapshot, this);
+            if ((object != null) && (object.getSnapshotId() == null)) {
+                object.setSnapshotId(snapshotIdService.generateId());
+            }
+            if (resultInitialized && result != object) {
+                Scope.getCurrentScope().getLog(getClass())
```
Comment (fbiville) [inline liquibase-standard/src/main/java/liquibase/snapshot/SnapshotGeneratorChain.java:62]: I actually need to double-check this, not sure it makes sense

File: liquibase-standard/src/test/groovy/liquibase/snapshot/jvm/ViewSnapshotGeneratorTest.groovy [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import liquibase.database.Database
+import liquibase.snapshot.DatabaseSnapshot
+import liquibase.snapshot.SnapshotControl
+import liquibase.structure.core.Schema
+import liquibase.structure.core.View
+import spock.lang.Specification
+
+class ViewSnapshotGeneratorTest extends Specification {
+
+    def "does not add any views to snapshot if database does not support views"() {
```
Comment (fbiville) [inline liquibase-standard/src/test/groovy/liquibase/snapshot/jvm/ViewSnapshotGeneratorTest.groovy:12]: probably need to update this or add a new test with view included but unsupported by db

File: liquibase-standard/src/main/java/liquibase/snapshot/SnapshotGeneratorChain.java (hunk 5 above)
Comment (fbiville) [reply liquibase-standard/src/main/java/liquibase/snapshot/SnapshotGeneratorChain.java:62]: I was also tempted to throw here
ideally subsequent snapshot generators should only modify instance of database objects (mostly by adding nested attributes), not return new ones

Comment (fbiville) [reply liquibase-standard/src/main/java/liquibase/snapshot/SnapshotGeneratorChain.java:62]: ok, I tried the exception approach and it works out quite well in my case. If I forget to replace the built-in catalog snapshot generator, I get this error:

```java
liquibase.exception.DatabaseException: Snapshot generator liquibase.snapshot.jvm.CatalogSnapshotGenerator has returned a different reference from the previous generator liquibase.ext.neo4j.snapshot.TypeSnapshotGeneratorNeo4j.
	Snapshot object was: class liquibase.structure.core.Catalog@1202508091, it is now: class liquibase.structure.core.Catalog@1871617286.
	Consider adding one of the generator to the result of the other's implementation of liquibase.snapshot.SnapshotGenerator#replaces.
```

Comment (fbiville) [reply liquibase-standard/src/main/java/liquibase/snapshot/SnapshotGeneratorChain.java]: added in https://github.com/liquibase/liquibase/pull/5619/commits/e4fcf69684b3738c02c608e525594ed799a3782e

File: liquibase-standard/src/main/java/liquibase/database/core/HsqlDatabase.java [hunk 7]
```diff
@@ -380,6 +382,21 @@ public boolean supportsInitiallyDeferrableColumns() {
         return false;
     }
 
+    @Override
+    public boolean supports(Class<? extends DatabaseObject> object) {
+        if (Schema.class.isAssignableFrom(object)) {
+            return false;
```
Comment (fbiville) [inline liquibase-standard/src/main/java/liquibase/database/core/HsqlDatabase.java:388]: oopsie, that is 100% wrong

Comment (fbiville) [reply liquibase-standard/src/main/java/liquibase/database/core/HsqlDatabase.java]: fixed in 775f9124bb727253c07dc9f10babc4398ee7caeb

File: liquibase-standard/src/main/java/liquibase/snapshot/SnapshotControl.java (hunk 4 above)
Comment (fbiville) [reply liquibase-standard/src/main/java/liquibase/snapshot/SnapshotControl.java:127]: Note: I had to update `liquibase.diff.output.changelog.DiffToChangeLogTest#getOrderedOutputTypes_hasDependencies` as a consequence

**Comment by filipelautert** [comment]
@fbiville seems ready to go!

**Review by filipelautert — APPROVED** [review]

**Comment by fbiville** [comment]
> There is a merge conflict with `SpringLiquibase` class, do you mind resolving it? Or I can do it if you wish.

PR is now rebased and the conflict is fixed.

**Comment by rberezen** [comment]
@fbiville hello! I see that some of our functional tests have failed during diff, generateChangelog, etc. commands with:

```
java.lang.AssertionError: liquibase.exception.DatabaseException: Snapshot generator com.datical.liquibase.ext.storedlogic.function.mysql.MySQLFunctionSnapshotGenerator has returned a different reference from the previous generator com.datical.liquibase.ext.storedlogic.function.FunctionSnapshotGenerator.
	Snapshot object was: class com.datical.liquibase.ext.storedlogic.function.Function@709531076, it is now: class com.datical.liquibase.ext.storedlogic.function.Function@119202251.
	Consider declaring com.datical.liquibase.ext.storedlogic.function.mysql.MySQLFunctionSnapshotGenerator as being replaced by one the generator in the chain via liquibase.snapshot.SnapshotGenerator#replaces.
	at org.junit.Assert.fail(Assert.java:88)
	at liquibase.harness.util.TestUtils.executeCommandScope(TestUtils.groovy:67)
	at liquibase.harness.change.ChangeObjectTests.apply #testInput.changeObject against #testInput.databaseName #testInput.version(ChangeObjectTests.groovy:79)
```
Could you please take a look when you get a chance? Let me know if you need any extra information. 

https://github.com/liquibase/liquibase-test-harness/actions/runs/8272583142
https://github.com/liquibase/liquibase-pro-tests/actions/runs/8272584158

cc @filipelautert @MalloD12

**Review by rberezen — APPROVED** [review]

**Comment by jasonlyle88** [comment]
Hello @fbiville and @filipelautert ,

As of this PR, I am no longer able to run `mvn clean install` on the master branch of the project. I did some testing, and the commit right before this (`58e9c3d09c3db57cd149bb27dba19539327248d8`) allows me to run `mvn clean install` just fine on the project. As of this commit (`2b1c363e7f998bcdb1e4a6c66906a0e2c2226f77`), it no longer works.

When running `mvn clean install` I get an error on `liquibase-integration-tests` for the hsqldb database type. Deleting the `liquibase-integration-tests/src/test/java/liquibase/dbtest/hsqldb/HsqlIntegrationTest.java` file allows `mvn clean install` to run correctly... but obviously that isn't a real solution. Attached is the [liquibase.dbtest.hsqldb.HsqlIntegrationTest.txt](https://github.com/liquibase/liquibase/files/15201113/liquibase.dbtest.hsqldb.HsqlIntegrationTest.txt) file that is generated and contains the failures. Below is also the end of the output of the maven run if it is also helpful:

```
[INFO]
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testClearChecksums:843 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/hsqldb/complete/root.changelog.xml::1.1::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQ_PERSON in statement [CREATE SEQUENCE PUBLIC.seq_person AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seq_person AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testDbDoc:926 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/hsqldb/complete/root.changelog.xml::1.1::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQ_PERSON in statement [CREATE SEQUENCE PUBLIC.seq_person AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seq_person AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testDiff:638->AbstractIntegrationTest.runCompleteChangeLog:356->AbstractIntegrationTest.runChangeLogFile:368 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/hsqldb/complete/root.changelog.xml::1.1::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQ_PERSON in statement [CREATE SEQUENCE PUBLIC.seq_person AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seq_person AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testGenerateChangeLogWithNoChanges:1066->AbstractIntegrationTest.runCompleteChangeLog:356->AbstractIntegrationTest.runChangeLogFile:368 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/hsqldb/complete/root.changelog.xml::1.1::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQ_PERSON in statement [CREATE SEQUENCE PUBLIC.seq_person AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seq_person AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testRerunDiffChangeLog:660->AbstractIntegrationTest.runCompleteChangeLog:356->AbstractIntegrationTest.runChangeLogFile:368 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/hsqldb/complete/root.changelog.xml::1.1::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQ_PERSON in statement [CREATE SEQUENCE PUBLIC.seq_person AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seq_person AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testRollbackToChange:902 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/hsqldb/rollback/rollbackable.changelog.xml::8a::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQ_PUBLICATION_ID in statement [CREATE SEQUENCE PUBLIC.seq_publication_id AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seq_publication_id AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testRowCountPrecondition:1145->AbstractIntegrationTest.runChangeLogFile:368 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/common/common.tests.changelog.xml::seq-test::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQTESTONALL in statement [CREATE SEQUENCE PUBLIC.seqtestonall AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seqtestonall AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testRunUpdateOnOldChangelogTableFormat:423 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/hsqldb/complete/root.changelog.xml::1.1::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQ_PERSON in statement [CREATE SEQUENCE PUBLIC.seq_person AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seq_person AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testSnapshotReportsAllObjectTypes:329->AbstractIntegrationTest.runCompleteChangeLog:356->AbstractIntegrationTest.runChangeLogFile:368 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/hsqldb/complete/root.changelog.xml::1.1::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQ_PERSON in statement [CREATE SEQUENCE PUBLIC.seq_person AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seq_person AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testTableExistsPreconditionTableNameMatch:1111->AbstractIntegrationTest.runChangeLogFile:368 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/common/common.tests.changelog.xml::seq-test::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQTESTONALL in statement [CREATE SEQUENCE PUBLIC.seqtestonall AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seqtestonall AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testTableIsEmptyPrecondition:1121->AbstractIntegrationTest.runChangeLogFile:368 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/common/common.tests.changelog.xml::seq-test::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQTESTONALL in statement [CREATE SEQUENCE PUBLIC.seqtestonall AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seqtestonall AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testTag:628 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/hsqldb/complete/root.changelog.xml::1.1::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQ_PERSON in statement [CREATE SEQUENCE PUBLIC.seq_person AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seq_person AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testTagEmptyDatabase:861 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/hsqldb/rollback/rollbackable.changelog.xml::8a::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQ_PUBLICATION_ID in statement [CREATE SEQUENCE PUBLIC.seq_publication_id AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seq_publication_id AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testUpdateChangelogChecksum:1201 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/hsqldb/complete/included.changelog.xml::1.2::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQ_NEWS in statement [CREATE SEQUENCE PUBLIC.seq_news AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seq_news AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testUpdateClearUpdate:557 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/hsqldb/complete/root.changelog.xml::1.1::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQ_PERSON in statement [CREATE SEQUENCE PUBLIC.seq_person AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seq_person AS BIGINT]
[ERROR]   HsqlIntegrationTest>AbstractIntegrationTest.testUpdateTwice:543 » CommandExecution liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset changelogs/hsqldb/complete/root.changelog.xml::1.1::nvoxland:
     Reason: liquibase.exception.DatabaseException: object name already exists: SEQ_PERSON in statement [CREATE SEQUENCE PUBLIC.seq_person AS BIGINT] [Failed SQL: (-5504) CREATE SEQUENCE PUBLIC.seq_person AS BIGINT]
[INFO]
[ERROR] Tests run: 1137, Failures: 0, Errors: 16, Skipped: 602
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Liquibase Parent 0-SNAPSHOT:
[INFO]
[INFO] Liquibase Parent ................................... SUCCESS [ 42.811 s]
[INFO] liquibase-standard ................................. SUCCESS [02:00 min]
[INFO] liquibase-cli ...................................... SUCCESS [  9.873 s]
[INFO] liquibase-snowflake ................................ SUCCESS [  9.132 s]
[INFO] Liquibase .......................................... SUCCESS [  1.644 s]
[INFO] liquibase-maven-plugin ............................. SUCCESS [ 21.132 s]
[INFO] liquibase-cdi-jakarta .............................. SUCCESS [ 11.239 s]
[INFO] liquibase-cdi ...................................... SUCCESS [  9.956 s]
[INFO] liquibase-extension-testing ........................ SUCCESS [ 11.557 s]
[INFO] liquibase-integration-tests ........................ FAILURE [01:28 min]
[INFO] liquibase-dist ..................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  05:26 min
[INFO] Finished at: 2024-05-02T17:58:29-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.2.5:test (default-test) on project liquibase-integration-tests:
[ERROR]
[ERROR] Please refer to /Users/jlyle/temp/liquibase/liquibase-integration-tests/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :liquibase-integration-tests
```

Any thoughts on what is going on here?

**Comment by fbiville** [comment]
@jasonlyle88 I initially thought I missed porting `supportsSequences` to `supports(Class<? extends DatabaseObject>)` in `liquibase.database.core.HsqlDatabase`, but I don't think that's the issue.

`supports` is implemented like this:

```
    @Override
    public boolean supports(Class<? extends DatabaseObject> object) {
        if (Catalog.class.isAssignableFrom(object)) { // this is the same impl as supportsCatalog
            try {
                return getDatabaseMajorVersion() >= 2;
            } catch (DatabaseException e) {
                return true;
            }
        }
        return super.supports(object);
    }
```

This PR changed all `supportsSequences` calls to `supports(Sequence.class)`.
In that case, `super.supports` is going to be called, and it's going to return `true` which is 100% equivalent to calling `supportsSequences`. 🤔 

Let me dig deeper.

**Pull request merged**